File: 53.最大子数组和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子数组和
 */

// @lc code=start
func maxSubArray(nums []int) (res int) {
	pre := 0
	res = nums[0]
	for i := 0; i < len(nums); i++ {
		pre = max(nums[i], pre+nums[i]) // 相当于维持一个单调递增数组
		res = max(res, pre)
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// @lc code=end

File: 53.最大子数组和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子数组和
#

# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        length = len(nums)
        pre = 0
        res = nums[0]
        for i in range(length):
            pre = max(nums[i], nums[i] + pre)
            res = max(pre, res)
        return res
# @lc code=end

